src/NHibernate/Cfg/IConfigurationManager.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Configuration;
+using System.Linq;
+
+namespace NHibernate.Cfg
+{
+	internal static class ConfigurationManagerExtensions
+	{
+		//TODO 6.0:  Replace with GetAppSetting and document as possible breaking change all usages. 
+		internal static string GetAppSettingIgnoringCase(this IConfigurationManager config, string name)
+		{
+			if (!(config is SystemConfigurationManager))
+				return config.GetAppSetting(name);
+
+			var key = ConfigurationManager.AppSettings.Keys.Cast<string>().FirstOrDefault(k => name.Equals(k, StringComparison.OrdinalIgnoreCase));
+			return string.IsNullOrEmpty(key)
+				? null
+				: ConfigurationManager.AppSettings[key];
+		}
+	}
+
+	public interface IConfigurationManager
```

Comment:
I think it should be split into 2 or 3 different interfaces.

Reply:
Why? It seems to me all or nothing is required as with our tests.

Reply:
Because the ConfigurationManager was an implementation detail, and abstracting it away one-to-one does not sound as a right move. It would make more sense if we have 3 distinct interfaces: connection string provider, logger provider, and configuration provider.

Reply:
I disagree - this is abstraction for our exact "configuration from external source" needs. If user provides only one implementation - he can hit run-time `PlatformNotSupported` exception or something like that just because he didn't know that he must override three different interfaces at once.

User either uses some configuration file (app.config,xml,...) and then he must to override all current `IConfigurationManager` or don't need it at all  if he uses configuration by code.

I would think "why would I need to override Logger Provider" if I provide it explicitly by code? But in fact if "logger provider" is not overridden - runtime exception is thrown on Xamarin.

I was thinking to use `GetLoggerClass` instead  of `GetAppConfig` but decided to keep it just to be more flexible in future. But if you think we shouldn't provide general config access methods I can change it to `GetLoggerClass`. But I believe it should be single interface that covers "external configuration" scenario.

Reply:
I think the exception cases are orthogonal to the refactoring. I think we should deal with `ConfigurationErrorsException` as it is documented behaviour of `.AppSettings`/`.ConnectionStrings`/`GetSection`.

Reply:
These are 3 distinct cases. 

In case of logger we use an opportunistic approach (which I don't like) and it's OK if we just handle this exception. 

In case of connection string the user would not be able to use a named connection string and will need to use a provided connection string.

In case of configuration this is also configuration-over-default approach and we need to handle this exception.

Reply:
I don't get how you want to deal with exception and where? You suggest to implement three interfaces and silently ignore this exception right inside of each default implementation.

So for instance with our tests  - if only one provider is overridden all others will be silently ignored? That doesn't sound right to me.

Again I don't consider this interface as different service providers - it's exactly configuration provider. Now we expect 3 things from configuration and so it's OK to have 3 members and not 3 different interfaces.

This splitting looks like unnecessary complication to me. Not just in our code but and for end user.

Reply:
I've refactored code and got rid of App.config specific `GetAppConfig` method. Also instead of interface I introduced [abstract class](https://github.com/nhibernate/nhibernate-core/pull/2116/files#diff-db9ccbc60054326e75ad7f3a03212156R11-R20) (so new configuration settings can be added easily.)

So my main argument for this approach - all configuration related settings are present in one place and can be easily replaced by custom logic. IMHO that allows to fix both #1402 (not sure what you meant exactly there - but it's either already supported by using `SystemConfigurationManager` or can be easily implemented by end user) and #2105 (by making sure that no configuration calls are done)